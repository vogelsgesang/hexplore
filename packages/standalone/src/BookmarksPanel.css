.hv-bookmarks-panel {
    padding: .4em;
}
.hv-bookmarks-list {
    margin: .2em 0;
}

.hv-bookmarks-empty-placeholder {
    color: #555;
    text-align: center;
    margin: .5em 0;
}

.hv-bookmarks-list-entry {
    display: flex;
    margin: .2em 0;
}

.hv-bookmarks-list-name {
    flex: 1;
    min-width: 5em;
    margin-right: .2em;
}

.hv-bookmarks-hint {
    color: #555;
    text-align: center;
    border-top: 1px solid #dddd;
}

/*
.hexviewerconfigeditor {
    display: flex;
    flex-direction: column;
    height: 100%;
}

.hexviewerconfigeditor > .hv-renderer-list {
    flex-shrink: 1;
    min-height: 10em;
    overflow: auto;
    border-top: 1px solid rgba(0,0,0,0.25);
    border-bottom: 1px solid rgba(0,0,0,0.25);
}

.hexviewerconfigeditor > .hv-add-renderer {
    margin: .4em;
}

.hexviewerconfigeditor .hv-form-row {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    margin: .4em;
}

.hexviewerconfigeditor .hv-renderer-details {
    margin-left: 1em;
    border-left: .2em solid rgba(0,0,0,0.25);
}
*/