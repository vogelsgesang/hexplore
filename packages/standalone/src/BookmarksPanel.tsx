import React from "react";
import {HighlightRange} from "hexplore-hexview";
import ButtonGroup from "react-bootstrap/ButtonGroup";
import Button from "react-bootstrap/Button";
import produce from "immer";

import "./BookmarksPanel.css";

export interface Bookmark extends HighlightRange {
    name: string;
}

interface BookmarksPanelProps {
    bookmarks: Bookmark[];
    setBookmarks: (h: Bookmark[]) => void;
    goto: (b: Bookmark) => void;
}

export function BookmarksPanel({bookmarks, setBookmarks, goto}: BookmarksPanelProps) {
    function removeBookmark(pos: number) {
        setBookmarks(
            produce(bookmarks, draft => {
                draft.splice(pos, 1);
            }),
        );
    }
    function renameBookmark(pos: number, newName: string) {
        setBookmarks(
            produce(bookmarks, draft => {
                draft[pos].name = newName;
            }),
        );
    }

    const renderedBookmarks = [];
    for (let i = 0; i < bookmarks.length; ++i) {
        const bookmark = bookmarks[i];
        renderedBookmarks.push(
            <div key={bookmark.key} role="group" aria-label={bookmark.name} className="hv-bookmarks-list-entry">
                <input
                    aria-label="Bookmark name"
                    type="text"
                    className="hv-stealth-input hv-bookmarks-list-name"
                    value={bookmark.name}
                    onChange={e => renameBookmark(i, e.target.value)}
                />
                <ButtonGroup>
                    <Button onClick={() => goto(bookmark)} size="sm" variant="outline-primary">
                        Go to
                    </Button>
                    <Button
                        title="Remove"
                        onClick={removeBookmark.bind(undefined, i)}
                        size="sm"
                        variant="outline-danger"
                    >
                        X
                    </Button>
                </ButtonGroup>
            </div>,
        );
    }
    let bookmarksContent;
    if (renderedBookmarks.length > 0) {
        bookmarksContent = <div className="hv-bookmarks-list">{renderedBookmarks}</div>;
    } else {
        bookmarksContent = <div className="hv-bookmarks-empty-placeholder">No bookmarks, yet.</div>;
    }

    return (
        <div className="hv-bookmarks-panel">
            {bookmarksContent}
            <div className="hv-bookmarks-hint">
                Press <kbd>m</kbd> to create a bookmark for the currently selected range
            </div>
        </div>
    );
}
