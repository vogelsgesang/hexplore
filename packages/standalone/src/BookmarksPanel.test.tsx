import React from "react";
import snapshotRenderer from "react-test-renderer";
import {render, fireEvent, screen, getByTitle, getByText, getByLabelText} from "@testing-library/react";
import {BookmarksPanel, Bookmark} from "./BookmarksPanel";

const testBookmarkList: Bookmark[] = [
    {from: 0, to: 4, name: "Bookmark 1", key: "t1"},
    {from: 10, to: 20, name: "My 2nd bookmark", key: "t2"},
    {from: 30, to: 32, name: "Numero tres", key: "t3"},
];

const dummyProps = {
    setBookmarks: () => {},
    goto: () => {},
    exportRange: () => {},
};

test("displays an empty list correctly", () => {
    const component = snapshotRenderer.create(<BookmarksPanel {...dummyProps} bookmarks={[]} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("displays a bookmark list correctly", () => {
    const component = snapshotRenderer.create(<BookmarksPanel {...dummyProps} bookmarks={testBookmarkList} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("can remove a bookmark", () => {
    const cfgRef = {current: testBookmarkList};
    render(
        <BookmarksPanel
            {...dummyProps}
            bookmarks={testBookmarkList}
            setBookmarks={b => {
                cfgRef.current = b;
            }}
        />,
    );
    const group = screen.getByLabelText("My 2nd bookmark");
    fireEvent.click(getByTitle(group, /remove/i));
    expect(cfgRef.current.length).toBe(2);
    expect(cfgRef.current[0]).toBe(testBookmarkList[0]);
    expect(cfgRef.current[1]).toBe(testBookmarkList[2]);
});

test("can edit the name", () => {
    const cfgRef = {current: testBookmarkList};
    render(
        <BookmarksPanel
            {...dummyProps}
            bookmarks={testBookmarkList}
            setBookmarks={b => {
                cfgRef.current = b;
            }}
        />,
    );
    const group = screen.getByLabelText("My 2nd bookmark");
    fireEvent.change(getByLabelText(group, /bookmark name/i), {target: {value: "El segundo"}});
    expect(cfgRef.current.length).toBe(3);
    expect(cfgRef.current[0]).toBe(testBookmarkList[0]);
    expect(cfgRef.current[1].name).toBe("El segundo");
    expect(cfgRef.current[2]).toBe(testBookmarkList[2]);
});

test("can `go to` a bookmark", () => {
    const gotoFn = jest.fn();
    render(<BookmarksPanel {...dummyProps} bookmarks={testBookmarkList} goto={gotoFn} />);
    const group = screen.getByLabelText("My 2nd bookmark");
    fireEvent.click(getByText(group, /go to/i));
    expect(gotoFn).toHaveBeenCalledTimes(1);
    expect(gotoFn).toHaveBeenCalledWith(testBookmarkList[1]);
});

test("can `export` the range of a bookmark", () => {
    const gotoFn = jest.fn();
    render(<BookmarksPanel {...dummyProps} bookmarks={testBookmarkList} exportRange={gotoFn} />);
    const group = screen.getByLabelText("My 2nd bookmark");
    fireEvent.click(getByTitle(group, /Export/i));
    expect(gotoFn).toHaveBeenCalledTimes(1);
    expect(gotoFn).toHaveBeenCalledWith(testBookmarkList[1]);
});
