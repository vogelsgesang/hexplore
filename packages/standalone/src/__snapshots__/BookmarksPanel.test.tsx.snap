// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a bookmark list correctly 1`] = `
<div
  className="hv-bookmarks-panel"
>
  <div
    className="hv-bookmarks-list"
  >
    <div
      aria-label="Bookmark 1"
      className="hv-bookmarks-list-entry"
      role="group"
    >
      <input
        aria-label="Bookmark name"
        className="hv-stealth-input hv-bookmarks-list-name"
        onChange={[Function]}
        type="text"
        value="Bookmark 1"
      />
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-outline-primary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Go to
        </button>
        <button
          className="btn btn-outline-danger btn-sm"
          disabled={false}
          onClick={[Function]}
          title="Remove"
          type="button"
        >
          X
        </button>
      </div>
    </div>
    <div
      aria-label="My 2nd bookmark"
      className="hv-bookmarks-list-entry"
      role="group"
    >
      <input
        aria-label="Bookmark name"
        className="hv-stealth-input hv-bookmarks-list-name"
        onChange={[Function]}
        type="text"
        value="My 2nd bookmark"
      />
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-outline-primary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Go to
        </button>
        <button
          className="btn btn-outline-danger btn-sm"
          disabled={false}
          onClick={[Function]}
          title="Remove"
          type="button"
        >
          X
        </button>
      </div>
    </div>
    <div
      aria-label="Numero tres"
      className="hv-bookmarks-list-entry"
      role="group"
    >
      <input
        aria-label="Bookmark name"
        className="hv-stealth-input hv-bookmarks-list-name"
        onChange={[Function]}
        type="text"
        value="Numero tres"
      />
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-outline-primary btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Go to
        </button>
        <button
          className="btn btn-outline-danger btn-sm"
          disabled={false}
          onClick={[Function]}
          title="Remove"
          type="button"
        >
          X
        </button>
      </div>
    </div>
  </div>
  <div
    className="hv-bookmarks-hint"
  >
    Press 'm' to create a bookmark for the currently selected range
  </div>
</div>
`;

exports[`displays an empty list correctly 1`] = `
<div
  className="hv-bookmarks-panel"
>
  <div
    className="hv-bookmarks-empty-placeholder"
  >
    No bookmarks, yet.
  </div>
  <div
    className="hv-bookmarks-hint"
  >
    Press 'm' to create a bookmark for the currently selected range
  </div>
</div>
`;
